# Module 1: Introduction to Exploratory Analysis 

## Step 2:  Data Cleaning

A big part of exploratory analysis involves data validation. This means gaining an understanding of the distribution of data, any potential data issues that can affect the quality of your model. As we learnt during our theory lesson, "Data preparation accounts for about 80% of the work of data scientists.” [source](https://www.forbes.com/sites/gilpress/2016/03/23/data-preparation-most-time-consuming-least-enjoyable-data-science-task-survey-says/#27df67b36f63).

What we'll be doing in this notebook:
-----

 1.  Checking variable type
 2.  Checking for missing variables 
 3.  Number of observations in the dataset
 4.  Descriptive statistics

### Import packages

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
from datetime import datetime
import dateutil.parser

# The command below means that the output of multiple commands in a cell will be output at once
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

# The command below tells jupyter to display up to 80 columns, this keeps everything visible
pd.set_option('display.max_columns', 80)
pd.set_option('expand_frame_repr', True)

### Import dataset

We read in our merged dataset below. Don't forget to update the name with your own! 

In [2]:
data_path = '../data/'
df = pd.read_csv(data_path+'df.csv', low_memory=False)

In the cell below, we take a random sample of 2 rows to get a feel for the data.

In [3]:
df.sample(2)

Unnamed: 0  Unnamed: 0.1  Unnamed: 0.1.1       activity  \
95810       105444        105444          105386  Used Clothing   
112473      122115        122115          122057  General Store   

        basket_amount bonus_credit_eligibility  \
95810             NaN                    False   
112473            NaN                     True   

                                                borrowers  \
95810   [{'first_name': 'Samson', 'last_name': '', 'ge...   
112473  [{'first_name': 'Joshua', 'last_name': '', 'ge...   

        currency_exchange_loss_amount description.languages  \
95810                             NaN                ['en']   
112473                            NaN                ['en']   

                                     description.texts.en  \
95810   Samson is 57 years old, married, and has six c...   
112473  Joshua N. K. is twenty two years old, married ...   

       description.texts.es description.texts.fr description.texts.ru  \
95810                   NaN                  NaN                  NaN   
112473                  NaN                  NaN                  NaN   

        funded_amount funded_date      id  image.id  image.template_id  \
95810             550  2011-12-30  357392    917362                  1   
112473            275  2010-04-05  187168    511219                  1   

        journal_totals.bulkEntries  journal_totals.entries  lender_count  \
95810                            0                       0            18   
112473                           0                       0             3   

        loan_amount location.country location.country_code location.geo.level  \
95810           550            Kenya                    KE            country   
112473          275            Kenya                    KE               town   

       location.geo.pairs location.geo.type location.town    name  partner_id  \
95810                1 38             point           NaN  Samson       142.0   
112473               1 38             point         Thika  Joshua       138.0   

       payments planned_expiration_date posted_date    sector  status tags  \
95810        []                     NaN  2011-12-01  Clothing  funded   []   
112473       []                     NaN  2010-04-01    Retail  funded   []   

        terms.disbursal_amount terms.disbursal_currency  terms.disbursal_date  \
95810                  50000.0                      KES  2011-11-11T08:00:00Z   
112473                 20000.0                      KES  2010-03-03T08:00:00Z   

        terms.loan_amount     ...      themes     translator.byline  \
95810                 550     ...         NaN           Brian Arens   
112473                275     ...         NaN  Jacqueline Pittenger   

        translator.image                            use video.id  \
95810                NaN   to buy more clothes to sell.      NaN   
112473          278831.0  Flour, sugar, rice, soap etc.      NaN   

        video.thumbnailImageId video.title video.youtubeId borrower_count  \
95810                      NaN         NaN             NaN              1   
112473                     NaN         NaN             NaN              1   

        partner_average_loan_size_percent_per_capita_income  \
95810                                                54.8     
112473                                               40.1     

       partner_charges_fees_and_interest  \
95810                               True   
112473                              True   

                                        partner_countries  \
95810   [{'iso_code': 'KE', 'region': 'Africa', 'name'...   
112473  [{'iso_code': 'KE', 'region': 'Africa', 'name'...   

        partner_currency_exchange_loss_rate partner_default_rate  \
95810                              0.024923             0.619151   
112473                             0.214680             1.483890   

       partner_default_rate_note  partner_delinquency_rate  \
95810                        NaN 

### 1) Type Checking
<a id='type_check'></a>

Type is very important in Python programing, because it affects the types of functions you can apply to a series. There are a few different types of data you will see regularly (see [this](https://en.wikibooks.org/wiki/Python_Programming/Data_Types) link for more detail):
* **int** - a number with no decimal places. example: loan_amount field
* **float** - a number with decimal places. example: partner_id field
* **str** - str is short for string. This type formally defined as a sequence of unicode characters. More simply, string means that the data is treated as a word, not a number. example: sector
* **boolean** - can only be True or False. There is not currently an example in the data, but we will be creating a gender field shortly.
* **datetime** - values meant to hold time data. Example: posted_date

Let's check the type of our variables using the examples we saw in the cell above.

In [4]:
## Select variables by name
type_example = df[['loan_amount','partner_id', 'sector','posted_date']]
## Pull (3) random rows
type_example.sample(3)

loan_amount  partner_id       sector posted_date
41231          275       133.0  Agriculture  2015-05-29
50591         1225       202.0  Agriculture  2014-12-01
60307          350       164.0         Food  2014-03-16

In [5]:
## Check the first cell for a column
df['posted_datetime'].head(1)
## Check the datatype for a single column
df['posted_datetime'].dtype

0    2017-05-09 00:40:03
Name: posted_datetime, dtype: object

dtype('O')

Datatype 'O' is for object.

### 2) Do I have missing values?

<a id='missing_check'></a>

If we have missing data, is the missing data at random or not at random? If data is missing at random, the data distribution is still representative of the population. You can probably ignore the missing values as an inconvienience. However, if the data is missing systematically, any modeling you do may be biased. You should carefully consider the best way to clean the data, it may involve dropping some data. See [here](https://en.wikipedia.org/wiki/Missing_data) for additional information.

We want to see how many values are missing in certain variable columns. One way to do this is to count the number of null observations. 

For this, we wrote a short function to apply to the dataframe. 

We print out the first few observations, but you can remove the .head() to print out all columns. 

In [29]:
#Create a new function:
def num_missing(x):
  return sum(x.isnull())

#Applying per column:
print("Missing values per column:")
## Check how many are missing by column, and then check which ones have any missing values
print(df.apply(num_missing, axis=0).where(lambda x : x != 0).dropna().head(20)) 
#axis=0 defines that function is to be applied on each column

Missing values per column:
basket_amount                                           117372.0
currency_exchange_loss_amount                           102626.0
description.texts.en                                      4328.0
description.texts.es                                    118313.0
description.texts.fr                                    118313.0
description.texts.ru                                    118312.0
funded_date                                               5612.0
location.town                                             7909.0
planned_expiration_date                                  24913.0
terms.loss_liability.currency_exchange_coverage_rate      3869.0
terms.repayment_interval                                117372.0
themes                                                   89302.0
translator.byline                                        35954.0
translator.image                                         55299.0
use                                                       4327.

In [28]:
#Applying per row:
print("\nMissing values per row:")
missing_by_row = df.apply(num_missing, axis=1) 
#axis=1 defines that function is to be applied on each row
print(missing_by_row.head()) 


Missing values per row:
0    15
1    15
2    14
3    15
4    16
dtype: int64


Remember when we used a left join to merge in the partner data. That means we could have null (missing) values in our partner_id field. We check this in the cell below and find that we have 9,642 loans that are missing. Let's investigate and try and understand whether the data is missing at random or systematically missing.

In [8]:
null_basket_amount = df.loc[df['basket_amount'].isnull()]
len(null_basket_amount)
null_basket_amount.head(2)

117372

Unnamed: 0  Unnamed: 0.1  Unnamed: 0.1.1                        activity  \
16          16            16              16  Primary/secondary school costs   
21          21            21              21                    Celebrations   

    basket_amount bonus_credit_eligibility  \
16            NaN                    False   
21            NaN                    False   

                                            borrowers  \
16  [{'first_name': 'Sally ', 'last_name': '', 'ge...   
21  [{'first_name': 'Naomi', 'last_name': '', 'gen...   

    currency_exchange_loss_amount description.languages  \
16                            NaN                ['en']   
21                            NaN                ['en']   

                                 description.texts.en description.texts.es  \
16  Sally is an ambitious woman from Bomet, a maiz...                  NaN   
21  Naomi is a single mother of 2 children and she...                  NaN   

   description.texts.fr description.texts.ru  funded_amount funded_date  \
16                  NaN                  NaN            150  2017-05-09   
21                  NaN                  NaN            100  2017-05-08   

         id  image.id  image.template_id  journal_totals.bulkEntries  \
16  1291449   2515878                  1                           0   
21  1291404   2515811                  1                           0   

    journal_totals.entries  lender_count  loan_amount location.country  \
16                       0             6          150            Kenya   
21                       0             4          100            Kenya   

   location.country_code location.geo.level location.geo.pairs  \
16                    KE               town               1 38   
21                    KE               town               1 38   

   location.geo.type location.town   name  partner_id payments  \
16             point         Bomet  Sally       156.0       []   
21             point         nyeri  Naomi       386.0       []   

   planned_expiration_date posted_date        sector  status  \
16              2017-06-07  2017-05-08     Education  funded   
21              2017-06-07  2017-05-08  Personal Use  funded   

                                                 tags  terms.disbursal_amount  \
16      [{'name': '#Parent'}, {'name': '#Schooling'}]                 15000.0   
21  [{'name': '#Animals'}, {'name': '#Parent'}, {'...                 10000.0   

   terms.disbursal_currency  terms.disbursal_date  terms.loan_amount  \
16                      KES  2017-05-28T07:00:00Z                150   
21                      KES  2017-04-03T07:00:00Z                100   

       ...      themes translator.byline  translator.image  \
16     ...         NaN               NaN               NaN   
21     ...         NaN   Cheryl Strecker         1412668.0   

                                                  use video.id  \
16               to pay school fees for her children.      NaN   
21  to buy a goat for a celebration during the Eas...      NaN   

    video.thumbnailImageId video.title video.youtubeId borrower_count  \
16                     NaN         NaN             NaN              1   
21                     NaN         NaN             NaN              1   

    partner_average_loan_size_percent_per_capita_income  \
16                                               49.6     
21                                                0.0     

   partner_charges_fees_and_interest  \
16                              True   
21                              True   

                                    partner_countries  \
16  [{'iso_code': 'KE', 'region': 'Africa', 'name'...   
21  [{'iso_code': 'KE', 'region': 'Africa', 'name'...   

    partner_currency_exchange_loss_rate partner_default_rate  \
16                             0.431935             2.575299   
21                             0.120642             0.000000   

   partner_default_rate_note  partner_delinquency_r

It seems that the number of loans peaked in 2014, and has decreased. All of the loans issued in the second half of 2015, 2016, 2017 have a partner. This tells us these values are not missing at random, they are more likely to occur in data before June 2015, and are most likely to occur in the first half of 2014. We should probably just drop data with the partner_id missing. We do so below using the notnull() function which checks if the field is populated.

In [17]:
df = df.loc[df['partner_id'].notnull()]

As a sanity check that our filtering was correctly done we check the new number of rows. It makes sense!

In [18]:
len(df.index)

118316

### 3) Sanity Checks
<a id='obs_check'></a>

**Does the dataset match what you expected to find?**
- is the range of values what you would expect. For example, are all loan_amounts above 0.
- do you have the number of rows you would expect
- is your data for the date range you would expect. For example, is there a strange year in the data like 1880.
- are there unexpected spikes when you plot the data over time


In the command below we find out the number of loans and number of columns by using the function shape. You can also use len(df.index) to find the number of rows.

In [10]:
print('There are %d observations and %d features' % (df.shape[0],df.shape[1]))

There are 118316 observations and 87 features


This tells us there are 118,316 observations and 84 features. We learnt in the theory lesson that each row is an observation and each column is a potential feature. 118,000 matches with what we expect based upon our conversations with Kiva. This is a very healthy sample size for applying machine learning algorithms.

### 4) Descriptive statistics of the dataset

<a id='desc_stats'></a>

In Module 1, we learned about mean, frequency and percentiles as a powerful way to understand the distribution of the data. If you are unfamiliar with these terms or need a refresher [this](https://www.mathsisfun.com/data/frequency-grouped-mean-median-mode.html) overview should be helpful. The "describe" command below provides key summary statistics for each numeric column.

In [11]:
df.describe()

Unnamed: 0   Unnamed: 0.1  Unnamed: 0.1.1  basket_amount  \
count  118316.000000  118316.000000   118316.000000     944.000000   
mean    64433.371843   64433.371843    64398.230239       0.185381   
std     38178.638548   38178.638548    38157.158427       2.145937   
min         0.000000       0.000000        0.000000       0.000000   
25%     29578.750000   29578.750000    29564.750000       0.000000   
50%     66559.500000   66559.500000    66517.500000       0.000000   
75%     98296.250000   98296.250000    98238.250000       0.000000   
max    127957.000000  127957.000000   127899.000000      25.000000   

       currency_exchange_loss_amount  funded_amount            id  \
count                   15690.000000  118316.000000  1.183160e+05   
mean                        7.017447     473.439771  7.096035e+05   
std                        15.655207     676.663803  3.412929e+05   
min                         0.010000       0.000000  2.510000e+02   
25%                         1.220000     250.000000  4.202852e+05   
50%                         3.490000     350.000000  6.975760e+05   
75%                         8.280000     600.000000  1.003961e+06   
max                      1285.510000   50000.000000  1.292273e+06   

           image.id  image.template_id  journal_totals.bulkEntries  \
count  1.183160e+05           118316.0                    118316.0   
mean   1.540915e+06                1.0                         0.0   
std    6.043467e+05                0.0                         0.0   
min    4.090000e+02                1.0                         0.0   
25%    1.013473e+06                1.0                         0.0   
50%    1.575774e+06                1.0                         0.0   
75%    2.052755e+06                1.0                         0.0   
max    2.516905e+06                1.0                         0.0   

       journal_totals.entries   lender_count    loan_amount     partner_id  \
count                118316.0  118316.000000  118316.000000  118316.000000   
mean                      0.0      14.608227     495.159784     165.037552   
std                       0.0      20.216517     704.079168      66.001909   
min                       0.0       0.000000      25.000000       6.000000   
25%                       0.0       7.000000     250.000000     133.000000   
50%                       0.0      11.000000     375.000000     156.000000   
75%                       0.0      18.000000     600.000000     164.000000   
max                       0.0    1589.000000   50000.000000     526.000000   

       terms.disbursal_amount  terms.loan_amount  \
count            1.183160e+05      118316.000000   
mean             4.178579e+04         495.159784   
std              4.474124e+04         704.079168   
min              2.500000e+01          25.000000   
25%              2.000000e+04         250.000000   
50%              3.000000e+04         375.000000   
75%              5.000000e+04         600.000000   
max              1.596948e+06       50000.000000   

       terms.loss_liability.currency_exchange_coverage_rate  \
count                                      114447.000000      
mean                                            0.122154      
std                                             0.041529      
min                                             0.100000      
25%                                             0.100000      
50%                                             0.100000      
75%                                             0.100000      
max                                             0.200000      

       terms.repayment_term  translator.image     video.id  \
count         118316.000000      6.301700e+04    76.000000   
mean              13.461079      1.165835e+06  1291.697368   
std                8.147379      7.036716e+05  1027.999559   
min                2.000000      2.392400e+04   150.000000   
25%               11.000000      5.186710e+05   487.750000   
50%           

In order to get the same summary statistics for categorical columns (string) we need to do a little data wrangling. The first line of code filters for all columns that are a data type object. As we know from before this means they are considered to be a string. The final row of code provides summary statistics for these character fields.

In [12]:
categorical = df.dtypes[df.dtypes == "object"].index
df[categorical].describe()

activity                                          borrowers  \
count    118316                                             118316   
unique      149                                              29987   
top     Farming  [{'first_name': 'Anonymous', 'last_name': '', ...   
freq      26243                                               2420   

       description.languages  \
count                 118316   
unique                     4   
top                   ['en']   
freq                  118306   

                                     description.texts.en  \
count                                              113988   
unique                                             113859   
top     Hello Kiva Community! <br /><br />Meet Jane, w...   
freq                                                    5   

                                     description.texts.es  \
count                                                   3   
unique                                                  3   
top     Beatrice is 51 years old, married and supports...   
freq                                                    1   

                                     description.texts.fr  \
count                                                   3   
unique                                                  3   
top     Teresia is 37 years old, married and together ...   
freq                                                    1   

                                     description.texts.ru funded_date  \
count                                                   4      112704   
unique                                                  4        3452   
top     Gerald is a married man. He has 2 children. He...  2016-03-08   
freq                                                    1         413   

       location.country location.country_code location.geo.level  \
count            118316                118316             118316   
unique                1                     1                  2   
top               Kenya                    KE               town   
freq             118316                118316             110407   

       location.geo.pairs location.geo.type location.town       name payments  \
count              118316            118316        110407     118316   118316   
unique                 47                 1          1249      16047        1   
top                  1 38             point        Likoni  Anonymous       []   
freq                77254            118316          5035       3680   118316   

       planned_expiration_date posted_date       sector  status    tags  \
count                    93403      118316       118316  118316  118316   
unique                    1860        3174           15       3    5426   
top                 2014-03-26  2014-02-24  Agriculture  funded      []   
freq                       621         621        45638  112704   65318   

       terms.disbursal_currency  terms.disbursal_date terms.local_payments  \
count                    118316                118316               118316   
unique                        2                  5413                  736   
top                         KES  2017-02-01T08:00:00Z                   []   
freq                     115076                  2498               117372   

       terms.loss_liability.currency_exchange terms.loss_liability.nonpayment  \
count                                  118316                          118316   
unique                                      3                               2   
top                                    shared                          lender   
freq                                   113783                          115735   

       terms.repayment_interval terms.scheduled_payments               themes  \
count                       944                   118316                29014   
unique                        3                      277                   27   
top                     Monthly                 

In the table above, there are 4 really useful fields: 

1) **count** - total number of fields populated (Not empty). 

2) **unique** - tells us how many different unique ways this field is populated. For example 4 in description.languages tells us there are 4 different language descriptions. 

3) **top** - tells us the most popular data point. For example, the top activity in this dataset is Farming which tells us most loans are in Farming.

4) **freq** - tells us that how frequent the most popular category is in our dataset. For example, 'en' (english) is the language almost all descriptions (description.languages) are written in (118,306 out of 118,316).

### Moving on

Next we move on to exploratory data analysis, where we will examine common plotting methods! 